From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Phoenix616 <mail@moep.tv>
Date: Sun, 24 Nov 2019 23:00:58 +0100
Subject: [PATCH] (Origami) Optimize inventory API item handling

Original code by Minebench, licensed under GNU General Public License v3.0
You can find the original code on https://github.com/Minebench/Origami

diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
index 396a4ae3d5a829eda78ef98561333aea300aa722..bbc23e2d4f1113f2fbeed4afc8824a822da5fda5 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftInventory.java
@@ -273,8 +273,6 @@ public class CraftInventory implements Inventory {
     }
 
     private int firstPartial(ItemStack item) {
-        ItemStack[] inventory = this.getStorageContents();
-        ItemStack filteredItem = CraftItemStack.asCraftCopy(item);
         if (item == null) {
             return -1;
         }
@@ -283,6 +281,10 @@ public class CraftInventory implements Inventory {
             if (cItem != null && cItem.getAmount() < cItem.getMaxStackSize() && cItem.isSimilar(filteredItem)) {
                 return i;
             }
+            // Elytra start - Optimize inventory API (moved down from before null check)
+            ItemStack[] inventory = getStorageContents();
+            ItemStack filteredItem = CraftItemStack.asCraftCopy(item);
+            // Elytra end
         }
         return -1;
     }
@@ -295,9 +297,10 @@ public class CraftInventory implements Inventory {
         /* TODO: some optimization
          *  - Create a 'firstPartial' with a 'fromIndex'
          *  - Record the lastPartial per Material
-         *  - Cache firstEmpty result
+         *  - Cache firstEmpty result // Implemented in Elytra
          */
 
+        int firstFree = -2; // Elytra - Cache firstEmpty result
         for (int i = 0; i < items.length; i++) {
             ItemStack item = items[i];
             while (true) {
@@ -307,7 +310,11 @@ public class CraftInventory implements Inventory {
                 // Drat! no partial stack
                 if (firstPartial == -1) {
                     // Find a free spot!
-                    int firstFree = this.firstEmpty();
+                    // Elytra start - Cache firstEmpty result
+                    if (firstFree == -2) {
+                        firstFree = firstEmpty();
+                    }
+                    // Elytra end
 
                     if (firstFree == -1) {
                         // No space at all!
@@ -320,9 +327,11 @@ public class CraftInventory implements Inventory {
                             stack.setAmount(this.getMaxItemStack());
                             this.setItem(firstFree, stack);
                             item.setAmount(item.getAmount() - this.getMaxItemStack());
+                            firstFree = -2; // Elytra - Cache firstEmpty result
                         } else {
                             // Just store it
                             this.setItem(firstFree, item);
+                            firstFree = -2; // Elytra - Cache firstEmpty result
                             break;
                         }
                     }
